### DISCUSSION TITLE
Function definitions stop at `INT3` instructions.

### DISCUSSION THREAD

**Discussion by HexPandaa** [NationalSecurityAgency/ghidra#5148] [Q&A]
**Describe the bug**
A function ends at an `INT3` instruction, even when it is `NOP`ed.

**To Reproduce**
1. Decompile a binary containing an `INT3` instruction.
2. Look at the range of the decompiled function.

**Expected behavior**
The function would either go to it's normal end, or 

**Screenshots**
Selecting the `INT3` instruction, the decompiler considers the cursor is in the function.
![image](https://user-images.githubusercontent.com/47880094/209235542-23f4ede7-16a0-452b-8153-199bd0bc3dbd.png)

Selecting the next instruction (`JMP`), the decompilation is the same, but the cursor is now in an undefined function.
![image](https://user-images.githubusercontent.com/47880094/209235785-9691b349-c697-4424-82d2-05a5409f2f75.png)

When patching the instruction as a `NOP`, the function isn't enlarged.
![image](https://user-images.githubusercontent.com/47880094/209236411-a02d056c-8642-42a9-90c9-974fd906f97e.png)
> Here we are considered to be in another (thunk) function that overlaps with the first one, but when printing the start and end address with a script, they did not change:
```bash
TestScript.java> thunk_FUN_08048063
TestScript.java> 	ram:08048060
TestScript.java> 	ram:08048061
TestScript.java> FUN_08048063
TestScript.java> 	ram:08048063
TestScript.java> 	ram:08048077
```

**Environment (please complete the following information):**
 - OS: Manjaro, kernel 6.0.11
 - Java Version: 19.0.1
 - Ghidra Version: 10.2.2
 - Ghidra Origin: AUR

**Additional context**
I'm working on improving the default `ResolveX86orX64LinuxSyscallsScript.java`. To do so, I'm working with an x86 32 bits Linux binary that's using inline syscalls to obfuscate it's behaviour. (here by setting up a signal handler and then triggering that signal via the `INT3` instruction). However, as the script (that I modified to support the `INT 0x80` syscall call) extract information only from defined functions, it does not find the `INT 0x80` instructions after the first `INT3`, even when the `INT3` is replaced with a `NOP`.
A way to define manually the function end, or to ignore these `INT3` instructions would be very practical. I tried setting up a `JUMP_OVERRIDE_UNCONDITIONNAL` to the next address on the `INT3` as I'm doing with syscalls (with a `CALL`), but the function range stayed the same.

**Comment by ghidracadabra** [comment]
If you replace an `INT3` with a `NOP`, you'll have to re-create the function in order for Ghidra's view of the function body to change.  You can do this by right-clicking on the function signature in the Listing and selecting Function -> Re-create Function.

The pcode for `INT3` contains a callother op, an indirect call, and then a return.  You could change the indirect call to a direct call by using a `CALL_OVERRIDE_UNCONDITIONAL` reference, but the return will still be there.  You can change the callother to a jump by using a `CALLOTHER_OVERRIDE_JUMP` reference.  This will effectively skip over the indirect call and the return.  You'll get a warning message, but I think it will work for your use case.

Alternatively, you could apply a fallthrough override on the instruction immediately before it (right-click on the previous instruction and select Fallthrough->Set) and `JUMP_OVERRIDE_UNCONDITIONAL` references on any jumps to the `INT3` instruction.  I'm not sure whether this would work better for you, and you'd still have to re-create the function, but I thought I'd mention it as a possibility.

**Comment by HexPandaa** [comment]
> Alternatively, you could apply a fallthrough override on the instruction immediately before it (right-click on the previous instruction and select Fallthrough->Set) and `JUMP_OVERRIDE_UNCONDITIONAL` references on any jumps to the `INT3` instruction. I'm not sure whether this would work better for you, and you'd still have to re-create the function, but I thought I'd mention it as a possibility.

I tried that option and it seems to be working quite well, thank you.
However I could not find a way to re-create the function via the exposed API, is there a way to do it in a script (it worked manually but my goal is to automate it)?

**Comment by ghidracadabra** [comment]
You can do this using [this constructor](https://github.com/NationalSecurityAgency/ghidra/blob/edc7515e292bb3f845b846c21b22a86f67b08714/Ghidra/Features/Base/src/main/java/ghidra/app/cmd/function/CreateFunctionCmd.java#L72) of the `CreateFunctionCmd` class.

For an example of how to use a command in a script, see [here](https://github.com/NationalSecurityAgency/ghidra/blob/edc7515e292bb3f845b846c21b22a86f67b08714/Ghidra/Features/Base/ghidra_scripts/DemangleAllScript.java#L54).
